lib/SILOptimizer/Transforms/AccessEnforcementDom.cpp

```diff
@@ -0,0 +1,257 @@
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "access-enforcement-dom"
+
+#include "swift/SIL/DebugUtils.h"
+#include "swift/SIL/MemAccessUtils.h"
+#include "swift/SIL/SILFunction.h"
+#include "swift/SILOptimizer/Analysis/DominanceAnalysis.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/Local.h"
+
+using namespace swift;
+
+namespace {
+class DominatedAccessRemoval {
+public:
+  using AccessedStoragePair = std::pair<BeginAccessInst *, AccessedStorage>;
+  using AccessedStorageInfo = llvm::SmallVector<AccessedStoragePair, 32>;
+  using DominatorToDominatedPair =
+      std::pair<BeginAccessInst *, BeginAccessInst *>;
+  using DomPairSet = llvm::SmallVector<DominatorToDominatedPair, 32>;
+  using KeyPathEntryPointsSet = llvm::SmallSet<SILInstruction *, 8>;
+
+public:
+  DominatedAccessRemoval(SILFunction &func, DominanceInfo *domInfo)
+      : func(func), domInfo(domInfo) {}
+
+  void perform();
+
+protected:
+  bool visitInstruction(SILInstruction *instr,
+                        AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+  void visitBeginAccess(BeginAccessInst *beginAccess, AccessedStorage storage,
+                        AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+  bool domByKeyPath(BeginAccessInst *dominatedInstr);
+  bool analyzeDomSubTree(SILBasicBlock *block,
+                         AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+
+  bool analyze();
+  void optimize();
+
+private:
+  SILFunction &func;
+  DominanceInfo *domInfo;
+  // domPairs is a vector of (Dominating begin_access, Dominated begin_access)
+  // both accesses are dynamic, and the Dominated one has no nested conflict
+  // we can turn the dominated to static during the optimize() phase
+  DomPairSet domPairs;
+  // All stdlib entry points that we've found in this function
+  KeyPathEntryPointsSet keypathEntries;
+};
+} // namespace
+
+bool DominatedAccessRemoval::domByKeyPath(BeginAccessInst *dominatedInstr) {
+  for (SILInstruction *keyPathEntry : keypathEntries) {
+    if (domInfo->properlyDominates(keyPathEntry, dominatedInstr)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool DominatedAccessRemoval::visitInstruction(
+    SILInstruction *instr,
+    AccessedStorageInfo &visitedDomAccessesToStorageInfo) {
+  if (auto *BAI = dyn_cast<BeginAccessInst>(instr)) {
+    if (BAI->getEnforcement() != SILAccessEnforcement::Dynamic) {
+      return true;
+    }
+    AccessedStorage storage = findAccessedStorageNonNested(BAI->getSource());
+    if (!storage) {
+      return true;
+    }
+    visitBeginAccess(BAI, storage, visitedDomAccessesToStorageInfo);
+  } else if (auto fullApply = FullApplySite::isa(instr)) {
+    SILFunction *callee = fullApply.getReferencedFunction();
+    if (!callee)
+      return true;
+    if (!callee->hasSemanticsAttr("_keyPathEntryPoint"))
```

Comment:
Is this the right check? Isn't it a decl attribute and not a semantic attribute?